usr.sbin/inetd/parse_v2.c

```diff
@@ -0,0 +1,1067 @@
+static hresult	bind_handler(struct servtab *, vlist);
+static hresult	exec_handler(struct servtab *, vlist);
+static hresult	filter_handler(struct servtab *, vlist);
+static hresult	group_handler(struct servtab *, vlist);
+static hresult	service_max_handler(struct servtab *, vlist);
+static hresult  ip_max_handler(struct servtab *, vlist);
+static hresult	protocol_handler(struct servtab *, vlist);
+static hresult	recv_buf_handler(struct servtab *, vlist);
+static hresult	send_buf_handler(struct servtab *, vlist);
+static hresult	socket_type_handler(struct servtab *, vlist);
+static hresult	unknown_handler(struct servtab *, vlist);
+static hresult	user_handler(struct servtab *, vlist);
+static hresult	wait_handler(struct servtab *, vlist);
+
+#ifdef IPSEC
+static hresult	ipsec_handler(struct servtab *, vlist);
+#endif
+
+static invoke_result	parse_invoke_handler(bool *, char **, struct servtab *);
+static bool fill_default_values(struct servtab *);
+static bool parse_quotes(char **);
+static bool	skip_whitespace(char **);
+static int	size_to_bytes(char *arg);
+static bool infer_protocol_ip_version(struct servtab *);
+static bool	setup_internal(struct servtab *);
+static void	try_infer_socktype(struct servtab *);
+#ifdef IPSEC
+static void	setup_ipsec(struct servtab *);
+#endif
+
+/* v2 Key handlers infrastructure */
+
+/* v2 syntax Handler function, which must parse all values for its key */
+typedef hresult (*key_handler_func)(struct servtab *, vlist);
+
+/* List of v2 syntax key handlers */
+static struct key_handler {
+	const char *name;
+	key_handler_func handler;
+} key_handlers[] = {
+	{"bind", bind_handler},
+	{"socktype", socket_type_handler},
+	{"acceptfilter", filter_handler},
+	{"protocol", protocol_handler},
+	{"sndbuf", send_buf_handler},
+	{"recvbuf", recv_buf_handler},
+	{"wait", wait_handler},
+	{"service_max", service_max_handler},
+	{"user", user_handler},
+	{"group", group_handler},
+	{"exec", exec_handler},
+	{"args", args_handler},
+	{"ip_max", ip_max_handler},
+#ifdef IPSEC
+	{"ipsec", ipsec_handler}
+#endif
+};
+
+#define STRINGIFY(x) #x
+#define TOSTRING(x) STRINGIFY(x)
+
+/* Error Not Initialized */
+#define ENI(key) ERR("Required option '%s' not specified", (key))
+
+#define WAIT_WRN "Option 'wait' for internal service '%s' should be '%s'"
+
+/* Too Few Arguemnts (values) */
+#define TFA(key) ERR("Option '%s' has too few arguments", (key))
+
+/* Too Many Arguments (values) */
+#define TMA(key) ERR("Option '%s' has too many arguments", (key))
+
+/* Too Many Definitions */
+#define TMD(key) ERR("Option '%s' is already specified", (key))
+
+#define VALID_SOCKET_TYPES "stream, dgram, rdm, seqpacket, raw"
+
+parse_v2_result
+parse_syntax_v2(struct servtab *sep, char **cpp)
+{
+
+	/* Catch multiple semantic errors instead of skipping after one */
+	bool is_valid_definition = true;
+	/* Line number of service for error logging. */
+	size_t line_number_start = line_number;
+	invoke_result result;
+
+	if (debug) {
```

Comment:
Perhaps use a DPRINTF() macro to allow compiling out the debugging code.

Reply:
The debugging code is enabled using the -d flag at runtime. Wouldn't that sort of negate the need for a macro? I think a DPRINTF macro would still be helpful and allow disabling of debugging for someone compiling the program who doesn't want it, but its purpose is mainly to check that a config is loaded correctly.

Reply:
I was thinking that a macro is less clutter, and can be compiled away if needed. So it has dual benefit.